api/v1alpha1/jvmchaos_webhook.go

```diff
@@ -76,3 +85,557 @@ func (in *JVMChaos) ValidateScheduler(spec *field.Path) field.ErrorList {
+
+		if actionPR, actionOK := actions[in.Spec.Action]; actionOK {
+			if actionPR.Flags != nil {
+				allErrs = append(allErrs, in.validateParameterRules(in.Spec.Flags, actionPR.Flags, flagsField, targetField, actionField)...)
+			}
+
+			if actionPR.Matchers != nil {
+				allErrs = append(allErrs, in.validateParameterRules(in.Spec.Matchers, actionPR.Matchers, matcherField, targetField, actionField)...)
+			}
+
+		} else {
+			supportActions := make([]JVMChaosAction, 0)
+			for k := range actions {
+				supportActions = append(supportActions, k)
+			}
+
+			notSupportedError := field.NotSupported(actionField, in.Spec.Action, toString(supportActions))
+			errorMsg := fmt.Sprintf("target: %s does not match action: %s, action detail error: %s",
+				in.Spec.Target, in.Spec.Action, notSupportedError)
+			allErrs = append(allErrs, field.Invalid(targetField, in.Spec.Target, errorMsg))
+		}
+	} else {
+		allErrs = append(allErrs, field.Invalid(targetField, in.Spec.Target, "unknown JVM chaos target"))
+	}
+
+	return allErrs
+}
+
+func toString(actions []JVMChaosAction) []string {
+	ret := make([]string, 0)
+	for _, act := range actions {
+		ret = append(ret, string(act))
+	}
+	return ret
+}
+
+func (in *JVMChaos) validateParameterRules(parameters map[string]string, rules []ParameterRules, parent *field.Path, target *field.Path, action *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, rule := range rules {
+		innerField := parent.Child(rule.Name)
+
+		var value = ""
+		var exist = false
+		if parameters != nil {
+			value, exist = parameters[rule.Name]
+		}
+		if rule.Required && !exist {
+			errorMsg := fmt.Sprintf("with %s: %s, %s: %s", target, in.Spec.Target, action, in.Spec.Action)
+			allErrs = append(allErrs, field.Required(innerField, errorMsg))
+		}
+
+		if exist && rule.Required && rule.ParameterType == StringType {
+			if len(value) == 0 {
+				errorMsg := fmt.Sprintf("%s:%s cannot be empty", innerField, value)
+				allErrs = append(allErrs, field.Invalid(innerField, value, errorMsg))
+			}
+		}
+
+		if exist && rule.ParameterType == IntType {
+			_, err := strconv.Atoi(value)
+			if err != nil {
+				errorMsg := fmt.Sprintf("%s:%s cannot parse as Int", innerField, value)
+				allErrs = append(allErrs, field.Invalid(innerField, value, errorMsg))
+			}
+		}
+
+		if exist && rule.ParameterType == BoolType {
+			_, err := strconv.ParseBool(value)
+			if err != nil {
+				errorMsg := fmt.Sprintf("%s:%s cannot parse as boolean", innerField, value)
+				allErrs = append(allErrs, field.Invalid(innerField, value, errorMsg))
+			}
+		}
+	}
+	return allErrs
+}
+
+// +kubebuilder:object:generate=false
+
+// JvmSpec from chaosblade-jvm-spec.yaml
+// chaosblade-jvm-spec.yaml file generated by
+// https://github.com/chaosblade-io/chaosblade-exec-jvm/blob/master/chaosblade-exec-service/src/main/java/com/alibaba/chaosblade/exec/service/build/SpecMain.java
+var JvmSpec = map[JVMChaosTarget]map[JVMChaosAction]ActionParameterRules{
```

Comment:
Does `JvmSpec` need manual maintenance? Any idea to automate it?

Reply:
@STRRL 

Yes,if there is an update or version upgrade to the chaosBlade -jvm-spec.yaml file, we may need to update the JVMSpec. There are partial differences between the definition of JVMSpec and that of chaosBlade -jvm-spec.yaml. The main difference is in the type definition.  A clearer type definition would help us do some data validation ahead of time. The difference is artificial, and there may be some specific criteria. ChaosBlade -exec-JVM does not have frequent version updates, so I find it acceptable to maintain it manually for now.

Reply:
Oh, I got that.